app/controllers/users_controller.rb

```diff
@@ -1,17 +1,31 @@
 # frozen_string_literal: true
 
 class UsersController < ApplicationController
-  skip_before_action :check_token, only: %i[create create_session]
+  skip_before_action :check_token, only: %i[create confirm_email create_session]
 
   def create
     @user = User.create(create_params)
     if @user.save
       render status: :created, action: :show
+      UserMailer.email_confirmation(@user, @user.email).deliver_now
     else
       render status: :bad_request, json: @user.errors
     end
   end
 
+  def confirm_email
+    @user = User.find_by(email_confirmation_token: params[:email_confirmation_token])
+    if @user && !@user.activated
+      @user.update(activated: true)
+      render status: :ok, json: { success: 'Email confirmed successfully' }
+    elsif @user
+      @user.update(email: @user.unconfirmed_email, unconfirmed_email: nil)
+      render status: :ok, json: { success: 'Email updated successfully' }
+    else
+      render status: :bad_request, json: { error: 'Invalid request' }
+    end
+  end
+
```

Comment:
Needs to regenerate confirmation token in unconfirmed cases
Also could probably be slightly refactored to dedup logic